File: Sistemas Distribuídos/Protobuf/cliente.py
import socket

def main():
    HOST = "localhost"
    PORT = 5000
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((HOST, PORT))

    sair = True
    data = ""
    while bool(sair):
        opcao = int(input('''
        \n\n# MENU PRINCIPAL #\n
        1 - Criar contato
        2 - Procurar contato
        3 - Atualizar contato
        4 - Remover contato
        0 - Sair
        - > '''))
        print("Sua escolha foi:", opcao)
        if opcao == 1:
            nome = ''
            endereco = ''
            email = ''
            telefones = ''
            print("\nNOVO CONTATO\n") 
            nome = input('Digite o nome do novo contato: ')
            endereco = input('Digite o endereco do novo contato: ')
            email = input('Digite o email do novo contato: ')
            telefones = input('Digite o numero de telefones do novo contato (N telefones com "+" entre eles): ')
            
            contato = "1_" + nome +"_"+ endereco + "_"+ email + "_" + telefones + "\n"
            print("Mensagem que sera mandada para o server: " + contato)
            
            sock.sendto(contato.encode(),(HOST, PORT)) #enviando dados para o servidor
            data = sock.recv(2000).decode() #Recebendo dados do servidor
            
            if len(data) == 1: #solucao de um bug
                data = sock.recv(1024).decode()
            
            print("Mensagem recebida do servidor:", data)
            pass
        elif opcao == 2:
            print("\nPROCURAR CONTATO\n")
            nome = '' 
            nome = input('Digite o nome a ser procurado: ')
            contato = "2_"+ nome + "\n"
            sock.sendto(contato.encode(),(HOST, PORT)) #enviando para o servidor
            data = ''
            data = sock.recv(1024).decode()

            if len(data) == 1:
                data = sock.recv(1024).decode()
            print("Mensagem recebida do servidor:", data)
            data = ''
            pass
        elif opcao == 3: #UPDATE
            print("\nATUALIZAR CONTATO\n")
            nome = '' 
            nome = input('Digite o nome a ser atualizado: ')
            contato = "3_"+ nome + "\n"
            sock.sendto(contato.encode(),(HOST, PORT)) #enviando para o servidor
            data = ''
            data = sock.recv(1024).decode()

            if len(data) == 1:
                data = sock.recv(1024).decode()

            if data == 'Nao encontrei nenhum contato':
                print('Nao encontrei nenhum contato')
            else:
                print("Contato encontrado ---- ", data)
                contato2 = ''
                print("\nDADOS NOVOS DO CONTATO")
                nome = ''
                endereco = ''
                email = ''
                telefones = ''
                nome = input('Digite o nome do novo contato: ')
                endereco = input('Digite o endereco do novo contato: ')
                email = input('Digite o email do novo contato: ')
                telefones = input('Digite o numero de telefones do novo contato (N telefones com "+" entre eles): ')
                contato2 = "5_" + nome +"_"+ endereco +"_"+ email +"_"+ telefones + "\n"
                
                print("Msg que sera mandada para o server: " + contato2)
                sock.sendto(contato2.encode(),(HOST, PORT)) #enviando para o servidor
                
                data = ''
                data = sock.recv(1024).decode()

                if len(data) == 1:
                    data = sock.recv(1024).decode()

                print("Mensagem recebida do servidor:", data)
                data = ''
            pass
        elif opcao == 4: #DELETE
            print("\REMOVER CONTATO\n")
            nome = '' 
            nome = input('Digite o nome a ser removido: ')
            contato = "4_"+ nome + "\n"
            sock.sendto(contato.encode(),(HOST, PORT)) #enviando para o servidor
            data = ''
            data = sock.recv(1024)

            if len(data) == 1:
                data = sock.recv(1024).decode()
            print("Mensagem recebida do servidor:", data)
            data = ''
            pass
        elif opcao == 0:
            sock.sendto("0\n".encode(),(HOST, PORT))
            print("Encerrou a conexao com o server!")
            sock.close()
            sair = False
            pass
        else:
            print("\nEste número não está nas alternativas, tente novamente :D.")
        
        if len(data) == 1:
            data = sock.recv(1024).decode()
            data = ""
#-----------------------------------------------------
if __name__ == '__main__':
    main()

File: Sistemas Distribuídos/gRPC e RMI/RMI/src/main/java/rmi_crud/agenda/Cliente.java
package rmi_crud.agenda;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;


public class Cliente {
	
	public static void main (String[] args) throws Exception
	{
		InterfaceCrud stub = null;
		try
		{
			Registry registry = LocateRegistry.getRegistry(1099); 
			stub = (InterfaceCrud)registry.lookup("echoMsg");
		}
		catch (Exception e)
		{
			System.out.println("Falhou ao iniciar o Cliente.\n"+e);				
			System.exit(0);
		}
		
		String str; 
		boolean sair = false;
		BufferedReader kb = new BufferedReader(new InputStreamReader(System.in)); 

		while (sair==false) { 
			System.out.println("\n\n# MENU PRINCIPAL #\n");
			System.out.println("1 - Novo contato");
			System.out.println("2 - Procurar contato");
			System.out.println("3 - Atualizar contato");
			System.out.println("4 - Remover contato");
			System.out.println("0 - Sair");

			System.out.print("- > ");
			str = kb.readLine();

			if(str.equals("1")) { //CREATE

				String contato = "1_";

				System.out.println("\nNOVO CONTATO");

				System.out.print("Digite o nome: ");
				String nome = kb.readLine();

				contato += nome + "_";

				System.out.print("Digite o endereco: ");
				String endereco = kb.readLine();

				contato += endereco + "_";

				System.out.print("Digite o email: ");
				String email = kb.readLine();

				contato += email + "_";

				String telefones = "";

				while(true) {

					System.out.print("Digite o telefone (ou deixe em branco para sair): ");
					String telefone = kb.readLine();

					if(telefone.length()==0) {

						break;
					}

					telefones = telefones + telefone + "+";
				}
				System.out.println(telefones);
				telefones = telefones.substring(0, telefones.length()-1);
				System.out.println(telefones);
				contato += telefones;
				System.out.println(contato);
				contato += "\0";
				
				System.out.println("Recebido do servidor: " + stub.adicionarContato(contato));			
			}
            else if(str.equals("2")) { //READ
				String contato = "2_";
				System.out.println("\nPROCURAR CONTATO");
				System.out.print("Digite o nome a ser procurado: ");
				String nome = kb.readLine();
				contato += nome;
				System.out.println("Recebido do servidor: " + stub.procurarContato(contato));
			}
            else if(str.equals("3")) { // UPDATE
				String contato = "3_";
				System.out.println("\nATUALIZAR CONTATO");
				System.out.print("Digite o nome a ser atualizado: ");
				String nome = kb.readLine();
				contato += nome;
				
				String respostaServer = stub.atualizarContato(contato);
		        
		        if(respostaServer.equals("Nao encontrei nenhum contato")) {
					System.out.println(respostaServer);				
				}
				else {
					
					kb = new BufferedReader(new InputStreamReader(System.in));
					
					System.out.println("Recebi do servidor: "+ respostaServer);
					
					String contato2 = "5_";

					System.out.println("\nDADOS NOVOS DO CONTATO");

					System.out.print("Digite o nome: ");
					String nome2 = kb.readLine();

					contato2 += nome2 + "_";

					System.out.print("Digite o endereco: ");
					String endereco = kb.readLine();

					contato2 += endereco + "_";

					System.out.print("Digite o email: ");
					String email = kb.readLine();

					contato2 += email + "_";

					String telefones = "";

					while(true) {

						System.out.print("Digite o telefone (ou deixe em branco para sair): ");
						String telefone = kb.readLine();

						if(telefone.length()==0) {

							break;
						}

						telefones = telefones + telefone + "+";
					}
					//System.out.println(telefones);
					telefones = telefones.substring(0, telefones.length()-1);
					//System.out.println(telefones);
					contato2 += telefones;
					//System.out.println(contato);
					contato2 += "\0";
					System.out.println("ATUALIZAR: "+contato2);
					
					
					System.out.println("Recebido do servidor: " + stub.atualizarDadosContato(contato2));

				}
				
			}
            else if(str.equals("4")) { // DELETE
            	String contato = "4_";
				System.out.println("\nREMOVER CONTATO");
				System.out.print("Digite o nome a ser removido: ");
				String nome = kb.readLine();
				contato += nome;
				System.out.println("Recebido do servidor: " + stub.deletarContato(contato));
            }
		}
	}
}